drivers/ethernet/eth_xilinx_axi_ethernet_lite.c

```diff
@@ -0,0 +1,609 @@
+		size_t bytes_to_write_now = read_bytes + sizeof(uint32_t) > bytes_to_read
+						    ? bytes_to_read - read_bytes
+						    : sizeof(uint32_t);
+
+		ret += net_pkt_write(pkt, &current_data, bytes_to_write_now);
+
+		if (ret < 0) {
+			LOG_ERR("Write error bytes %zu/%zu (%zu)", read_bytes, bytes_to_read,
+				bytes_to_write_now);
+		} else {
+			LOG_DBG("Write OK bytes %zu/%zu (%zu) cursor %p remaining %d", read_bytes,
+				bytes_to_read, bytes_to_write_now, pkt->cursor.buf,
+				pkt->cursor.buf ? pkt->cursor.buf->size - pkt->cursor.buf->len : 0);
+		}
+
+		buffer_addr += sizeof(uint32_t);
+	}
+	return ret;
+}
+
+/* FIXME are there generic defines? */
+#define XILINX_AXI_ETHERNET_LITE_ARP_PACKET_LENGTH 28
+
+#define HEADER_BUF_SIZE                                                                            \
+	(sizeof(struct net_eth_hdr) + MAX(sizeof(struct net_ipv4_hdr), sizeof(struct net_ipv6_hdr)))
+#define HEADER_BUF_SIZE_ALIGNED                                                                    \
+	((HEADER_BUF_SIZE) + sizeof(uint32_t) - (HEADER_BUF_SIZE) % sizeof(uint32_t))
+
+static inline void
+xilinx_axi_ethernet_lite_receive(const struct xilinx_axi_ethernet_lite_config *config,
+				 struct xilinx_axi_ethernet_lite_data *data, mem_addr_t buffer_addr,
+				 mem_addr_t status_addr)
+{
+	size_t packet_size = NET_ETH_MTU;
+	uint8_t header_buf[HEADER_BUF_SIZE_ALIGNED];
+	uint16_t len;
+
+	struct net_pkt *pkt;
+	const struct net_eth_hdr *hdr;
+
+	if ((xilinx_axi_ethernet_lite_read_reg(config, status_addr) &
+	     XILINX_AXI_ETHERNET_LITE_RX_CTRL_READY_ENABLE_MASK) == 0) {
+		/* no data*/
+		return;
+	}
+
+	for (size_t read_bytes = 0; read_bytes < sizeof(header_buf);
+	     read_bytes += sizeof(uint32_t)) {
+		uint32_t current_data = xilinx_axi_ethernet_lite_read_reg(config, buffer_addr);
+
+		memcpy(&header_buf[read_bytes], &current_data, sizeof(current_data));
+		buffer_addr += sizeof(uint32_t);
+	}
+
+	hdr = (struct net_eth_hdr *)header_buf;
+	len = hdr->type;
+	/*
+	 * AXI Ethernet Lite cannot tell us the length of the received packet, so we try to parse it
+	 * Also, FCS is not used by Zephyr stack
+	 */
+	switch (ntohs(len)) {
+	case NET_ETH_PTYPE_ARP:
+		/* fixed length */
+		packet_size =
+			sizeof(struct net_eth_hdr) + XILINX_AXI_ETHERNET_LITE_ARP_PACKET_LENGTH;
+		break;
+	case NET_ETH_PTYPE_IP: {
+		const struct net_ipv4_hdr *ip4_hdr =
+			(const struct net_ipv4_hdr *)&header_buf[sizeof(*hdr)];
+		len = ip4_hdr->len;
+		packet_size = ntohs(len);
+		/* length includes ipv4 header length */
+		packet_size += sizeof(struct net_eth_hdr);
+		break;
+	}
+	case NET_ETH_PTYPE_IPV6: {
+		const struct net_ipv6_hdr *ip6_hdr =
+			(const struct net_ipv6_hdr *)&header_buf[sizeof(*hdr)];
+		/* payload + any optional extension headers */
+		len = ip6_hdr->len;
+		packet_size = ntohs(len);
+		packet_size += sizeof(struct net_eth_hdr) + sizeof(*ip6_hdr);
+		break;
+	}
+	default:
+		/* use the full MTU... */
+		break;
+	}
+
```

Comment:
the len of a ethernet packet is always at the same location, independent of independent of the protocol (ARP, ipv4, ipv6) in it.
use that.

https://docs.amd.com/r/en-US/pg135-axi-ethernetlite/Type/Length

Reply:
This type/length field is a part of the Ethernet MAC header.
This only indicates the packet length for the LLC (link-layer control) L2 protocol, a predecessor of modern Ethernet, which is not commonly used any more.
Modern networks generally use this field as a Type indicator (EtherType) instead.
For example, zephyr's network stack always interprets this as a *type* instead of a *length*, which is always the same for each protocol irrespective of the packet length.

For that reason, the IP has a separate *transmit size* MMIO register and does not rely on the type/length field for transmissions.
I have no idea why the same does not exist for RX; this was probably a design oversight.